### ISSUE TITLE
Integrating OpeNTF Variants into One Cohesive System

### ISSUE THREAD

**Issue by gabrielrueda** [fani-lab/OpeNTF#271]
# COMP-4960 Project - Integrating OpeNTF Variants into One Cohesive System

### Project Description
In OpeNTF there are other projects: Adila, GNN, and NMT. The idea is to be able to run these 3 subprojects all in the same project. i.e. from main.py as well as improve the modularity of the OpeNTF repository components.

### Main Tasks & Deadlines
- [x] Understanding/Reviewing How OpeNTF Runs and it's Parameters (January)
- [x] Understand the role and parameters for sub-projects (February)
    - [x] Adila 
    - [x] GNN
    - [x] NMT
    - [x] Curriculum-Based (CL) Loss
- [ ] Look into trying to utilize docker images (as in NMT project) (March)
- [ ] Finalize Design (May)
- [ ] Finish Code Implementation on a OpeNTF Branch (July)
- [ ] Final Report/Presentation (August)

**Comment by gabrielrueda** [comment]
@hosseinfani 
Update - February 9 2025

### OpeNTF (main branch)
- I have looked through the `run()` function in main.py, where I have investigated the different models that can be used as parameters as well as each models parent class (I looked at the class diagrams) 
- Main Models: random, fnn, bnn, fnn_emb, bnn_emb, nmt
- Those that start with "t" -> streaming scenario
- Those that end with "_a1" -> add 1 to the input
- Those that end with "_dt2v" -> doc to vector training

### GNN
- trains and evalutes graph neural networks
- from my understanding -> GNN will come up with embeddings that will then be used in a opeNTF model?
- GNN model options: gs, gin, gat, gatv2, gine

### Next Steps/Questions
- Can you confirm the list of sub-projects that we would be merging is correct?
- I will plan to meet with Jamil and Kap to learn more about the work they've done, and what I'd have to use from their projects. Is there anyone else I should meet with?

**Comment by hosseinfani** [comment]
Hi @gabrielrueda 

Very nice breakdown. Thank you.

Some more notes,
- For the NMT part, you can also talk with Kap @thangk
- For the OpeNTF part, we have different losses. A special one that needs attention is the Curriculum-based (CL) that plugs into when the loss will be calculated. For this, you need to contact me. But the paper is here: https://hosseinfani.github.io/res/papers/2025_WSDM_Adaptive_Loss-based_Curricula_for_Neural_Team_Recommendation.pdf and the codeline is in the cl-wsdm25 branch.

**Comment by hosseinfani** [comment]
Hi again @gabrielrueda 

Also, 
- I recall that we have developed a web app to show a leaderboard, showing the result of different methods in charts. You can also integrate that finally such that we have a visual interface to the library. This part is developed by @MarcoKurepa but unfortunately he passed away.
- if we could also the integration of containers/dockers and github workflow would be nice such that after each integration, we make sure that the pipeline build successfully.

**Comment by gabrielrueda** [comment]
@hosseinfani 
Update - February 22, 2025

I met with Kap and Jamil to learn about their projects and ideas towards merging the OpeNTF variants. 

When I met with Jamil, I learned the structure of how GNN works in opeNTF, where in the code it runs, and two types...
 there are two types:
	- gnn branch on the opeNTF repo: transfer learning...
		1. embeddings are generated by gnn
		2. those embeddings are inputted into the normal opeNTF process
	- the openGNTF repo -> does gnn directly generates the results

When I met with Kap, he suggested that each subproject should be containerized into it's own docker environment, since each repo has different requirements.  I looked at all the requirements and put them in a table highlighted below. 

We could try to compromise between the versioning (bring all variants to the highest version of the variants), but one comment I saw is that "opennmtpy requires 2.1 < torch < 2.3 and raises conflict with other versions when installed with pyg". This means that we probably cannot have GNN and NMT in the same environment. 

Another observation is that CL and the main branch have basically the same package versions. In terms of NMT and main branch, I think the package versions are close enough that we might be able to make it work.

We could probably have the following docker containers:
1. for main branch, NMT, and curriculum learning
2. for gnn (both variants) -> note: for the transfer learning GNN, we can have the embeddings generated on one container and then executed on the main branch container.


### Requirements:


| Package             | main branch | gnn branch (transfer learning to opeNTF) | openGNTF repo (all GNN) | nmt branch   | cl-wsdm25 |
| ------------------- | ----------- | ---------------------------------------- | ----------------------- | ------------ | --------- |
| torch               | >=1.9.0     | 2.5.0                                    | 2.3.0                   | 1.10.1+cu111 | >=1.9.0   |
| scipy               | 1.10.1      | 1.10.1                                   | 1.10.1                  | latest       | 1.10.1    |
| numpy               | 1.22.4      | 1.22.4                                   | 1.21.6                  | latest       | 1.22.4    |
| tqdm                | 4.65.0      | 4.65.0                                   | 4.66.2                  | latest       | 4.65.0    |
| pyyaml              | 6.0.1       | 6.0.1                                    | no                      | no           | 6.0.1     |
| scitkit-learn       | 1.2.2       | 1.2.2                                    | 1.3.2                   | latest       | 1.2.2     |
| pandas              | 1.5.3       | 1.5.3                                    | 2.0.3                   | latest       | 1.5.3     |
| gensim              | 4.3.1       | 4.3.1                                    | no                      | latest       | 4.3.1     |
| pytrec-eval-terrier | 0.5.2       | 0.5.2                                    | 0.5.6                   | no           | 0.5.2     |
| pytrec_eval         | no          | no                                       | no                      | latest       | no        |
| matplotlib          | 3.7.1       | 3.7.1                                    | no                      | latest       | 3.7.1     |
| opennmt-py          | 3.3         | no                                       | no                      | 3.0.4        | 3.3       |
| reranking           | 0.3.6       | 0.3.6                                    | no                      | no           | 0.3.6     |
| fairsearchcore      | 1.0.4       | 1.0.4                                    | no                      | no           | 1.0.4     |
| python-dateutil     | 2.8.2       | 2.8.2                                    | no                      | no           | 2.8.2     |
| bayesian-torch      | latest      | latest                                   | no                      | no           | no        |
| torchvision         | no          | 2.5.0                                    | 0.18.0                  | 0.11.2+cu111 | no        |
| torchaudio          | no          | 2.5.0<br>                                | 2.3.0                   | 0.10.1       | no        |
| torchmetrics        | no          | no                                       | 1.4.0                   | no           | no        |
| torch-geometic      | no          | latest                                   | 2.5.3                   | latest       | no        |
| pyg-lib             | no          | latest                                   | no                      | no           | no        |
| torch-scatter       | no          | latest                                   | 2.1.2                   | 2.0.9        | no        |
| torch-sparse        | no          | latest                                   | 0.6.18                  | 0.6.12       | no        |
| torch-cluster       | no          | latest                                   | 1.6.3                   | 1.6.0        | no        |
| torch-spline-conv   | no          | latest                                   | 1.2.2                   | 1.2.1<br>    | no        |
| pytz                | no          | no                                       | 2024.1                  | no           | no        |

### Next Steps
To see how to utilize the docker containers for each project, and then after that look into how to merge the CL codebase (since it seems like the most straight forward one to merge)

**Comment by hosseinfani** [comment]
Hi @gabrielrueda 
Thank you for the update and for the detailed report.

We need a pipeline following component-based software architecture. That is, when the user specifies a step in the pipeline (like preprocessing, training, ...) or pick a model (non-temporal, temporal, gnn-based,...), the required packages get installed dynamically and the required packages get involved dynamically. 

So, we can start with an environment that only needs the bare minimum, like python, scipy, ... Then, based on the chosen config or running setting, other required packages get installed and load to the pipeline. This means that the environment should be dynamically populated. I understand that this is not a common practice but an advanced one. 

Just an idea: let's have a requirement.txt or environment.yml file for minimum required packages. Then, for each specific setting that needs some packages, we can have different  requirement.txt or environment.yml files but they will be called upon for installation during the pipeline run.

Hope it helps.

@thangk @jamil2388

**Comment by hosseinfani** [comment]
@gabrielrueda 
for openmt-py, please double check with their repo. That comment is old and maybe there is no such requirement anymore.

**Comment by gabrielrueda** [comment]
Hi @hosseinfani 
Update - March 9, 2025

I have started investigating the nmt codebase, specifically the run() method in main and spotted where the key differences were. It seems that integrating nmt into the main branch shouldn't be too tricky. Mainly it's selecting the torch device (for cuda) needed for NMT and for nmt the output path is set in the nmt.py and not in main like in the other models.

For the opennmt-py requirement, I took at look at the repo, https://github.com/OpenNMT/OpenNMT-py, and it still says that pytorch v2 is needed, preferably v2.2.

For Curriculum Learning (CL), (branch: cl-wsdm25) I took through the paper and some of the codebase, specifically in main.py. From my understanding, it is a way to overcome the problem of recommending hard non-popular vs easy popular and to reduce popularity bias. There is one param for CL with 3 possible values: `SL` (non-parametric), `DP` (parametric) and `normal` (no curriculum learning).

Anyways, after looking at the main.py in the cl-wsdm25 branch, I didn't notice any difference between the main.py in the main branch and this branch. I was wondering if I needed to look elsewhere in the branch for the curriculum learning integration, or if it is already integrated into the main branch.

Finally, in terms of the high level architecture I have the following in mind so far based on what you've said:

Steps in Pipeline:
1. Preprocessing 
	1. converting raw data into sparse matrices
	2. (opt) transfer learning for GNN - to generate embeddings that can be used in any model
2. Training
	1. pick a model - like fnn, bnn, nmt, gnn (end to end version)
	2. set training params
3. Post-Processing
	1. Adila reranking

**Comment by hosseinfani** [comment]
Hi @gabrielrueda 
Thank you for the update.

Yes, everything about the nmt is handled in the nmt.py class that wraps the openmt. You're finding is correct.

--------------
Regarding the CL, it is used in fnn and bnn when in the learn() function when the training wanted to use a loss function. If you go to 
https://github.com/fani-lab/OpeNTF/blob/6067933ec461b1a2f1c94d1e173c32e4acb4a44e/src/mdl/fnn.py#L222
https://github.com/fani-lab/OpeNTF/blob/6067933ec461b1a2f1c94d1e173c32e4acb4a44e/src/mdl/fnn.py#L227
https://github.com/fani-lab/OpeNTF/blob/6067933ec461b1a2f1c94d1e173c32e4acb4a44e/src/mdl/fnn.py#L228

You'll see several if-else clauses based on what loss function should be used.  

Also, if we could come up with a pattern that the loss function would be injected in a nicer way ...

--------------
I agree with your proposed pipeline but what about 'test' and 'eval' steps. When the model is trained, we need to bring it to the test phase and apply it for test set. Also, after the test, we need to see the performance of models on the test set as evaluation.
